File: backend/app/schemas/analytics.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.enums import ModuleType


class SkillBandSummary(BaseModel):
    latest: float | None = None
    best: float | None = None
    average: float | None = None


class BandTrendPoint(BaseModel):
    attempt_id: UUID
    skill: ModuleType
    band_score: float
    attempted_at: datetime
    test_title: str
    version_number: int


class QuestionTypeAccuracy(BaseModel):
    question_type: str
    attempted: int = Field(ge=0)
    correct: int = Field(ge=0)
    incorrect: int = Field(ge=0)
    accuracy: float | None


class AnalyticsAttemptOption(BaseModel):
    attempt_id: UUID
    skill: ModuleType
    label: str
    band_score: float | None
    finished_at: datetime


class ContentTiming(BaseModel):
    attempt_id: UUID
    kind: str
    target_id: UUID
    active_seconds: int = Field(ge=0)


class LatestMockSummary(BaseModel):
    session_id: UUID
    test_title: str
    finished_at: datetime
    reading_band: float | None
    listening_band: float | None
    writing_band: float | None
    overall_band: float | None


class AnalyticsDashboard(BaseModel):
    total_finalized_attempts: int = Field(ge=0)
    total_active_seconds: int = Field(ge=0)
    average_attempt_seconds: float | None
    bands: dict[str, SkillBandSummary]
    completed_full_mocks: int = Field(ge=0)
    latest_project_overall: float | None
    latest_full_mock: LatestMockSummary | None
    trends: list[BandTrendPoint]
    question_types: list[QuestionTypeAccuracy]
    weak_areas: list[QuestionTypeAccuracy]
    attempts: list[AnalyticsAttemptOption]
    content_timing: list[ContentTiming]


class WritingCriteriaComparison(BaseModel):
    task1_overall: float | None = None
    task2_overall: float | None = None
    ta: float | None = None
    cc: float | None = None
    lr: float | None = None
    gra: float | None = None


class AttemptComparisonSide(BaseModel):
    attempt_id: UUID
    skill: ModuleType
    test_title: str
    version_number: int
    band_score: float | None
    raw_score: int | None
    max_score: int | None
    elapsed_seconds: int | None
    accuracy: float | None
    question_types: list[QuestionTypeAccuracy]
    writing: WritingCriteriaComparison | None = None


class AttemptComparison(BaseModel):
    same_skill: bool
    same_test_version: bool
    left: AttemptComparisonSide
    right: AttemptComparisonSide

File: frontend/src/lib/api/analytics.ts
import { z } from "zod";
import { apiRequest, type ApiRequester } from "./client";

const skillSchema = z.enum(["READING", "LISTENING", "WRITING"]);
const accuracySchema = z.object({ question_type: z.string(), attempted: z.number().int(), correct: z.number().int(), incorrect: z.number().int(), accuracy: z.number().nullable() });
const bandSummarySchema = z.object({ latest: z.number().nullable(), best: z.number().nullable(), average: z.number().nullable() });
export const analyticsSchema = z.object({
  total_finalized_attempts: z.number().int(), total_active_seconds: z.number().int(), average_attempt_seconds: z.number().nullable(),
  bands: z.record(z.string(), bandSummarySchema), completed_full_mocks: z.number().int(), latest_project_overall: z.number().nullable(),
  latest_full_mock: z.object({ session_id: z.string().uuid(), test_title: z.string(), finished_at: z.string(), reading_band: z.number().nullable(), listening_band: z.number().nullable(), writing_band: z.number().nullable(), overall_band: z.number().nullable() }).nullable(),
  trends: z.array(z.object({ attempt_id: z.string().uuid(), skill: skillSchema, band_score: z.number(), attempted_at: z.string(), test_title: z.string(), version_number: z.number().int() })),
  question_types: z.array(accuracySchema), weak_areas: z.array(accuracySchema),
  attempts: z.array(z.object({ attempt_id: z.string().uuid(), skill: skillSchema, label: z.string(), band_score: z.number().nullable(), finished_at: z.string() })),
  content_timing: z.array(z.object({ attempt_id: z.string().uuid(), kind: z.string(), target_id: z.string().uuid(), active_seconds: z.number().int() })),
});
const sideSchema = z.object({ attempt_id: z.string().uuid(), skill: skillSchema, test_title: z.string(), version_number: z.number().int(), band_score: z.number().nullable(), raw_score: z.number().int().nullable(), max_score: z.number().int().nullable(), elapsed_seconds: z.number().int().nullable(), accuracy: z.number().nullable(), question_types: z.array(accuracySchema), writing: z.object({ task1_overall: z.number().nullable(), task2_overall: z.number().nullable(), ta: z.number().nullable(), cc: z.number().nullable(), lr: z.number().nullable(), gra: z.number().nullable() }).nullable() });
const comparisonSchema = z.object({ same_skill: z.boolean(), same_test_version: z.boolean(), left: sideSchema, right: sideSchema });
export type AnalyticsDashboard = z.infer<typeof analyticsSchema>;
export type AttemptComparison = z.infer<typeof comparisonSchema>;

export async function getAnalytics(skill?: "READING" | "LISTENING" | "WRITING", request: ApiRequester = apiRequest): Promise<AnalyticsDashboard> {
  return analyticsSchema.parse(await request<unknown>(`/analytics${skill ? `?skill=${skill}` : ""}`));
}
export async function compareAttempts(left: string, right: string): Promise<AttemptComparison> {
  return comparisonSchema.parse(await apiRequest<unknown>(`/analytics/compare?left=${encodeURIComponent(left)}&right=${encodeURIComponent(right)}`));
}
